src/dynamicprompts/samplers/command_collection.py

```diff
@@ -0,0 +1,46 @@
+from __future__ import annotations
+
+from typing import Iterable
+
+from dynamicprompts.commands import Command
+from dynamicprompts.samplers.base import SamplerRouter
+from dynamicprompts.types import CommandList, StringGen
+
+
+class CommandCollection:
+    """
+    A class that holds a collection of commands and manages generating values for them.
+    """
```

Comment:
Instead of passing the router, you could pass the list of generators. 

The rationale behind the `CommandCollection` is that mixing finite and non finite samplers in the same prompt is not trivial. The `CombinatorialSampler` effectively combines all the Combinatorial commands in the prompt into one big sampler, e.g. `{!A|B} {!X|Y}` will generate:
```
A X
A Y
B X
B Y
```

The `CyclicalSampler` and `RandomSampler` treat each command independently. The `CommandCollection` extracts the random and cyclical command and stores them. They return their next value on request.  Cyclical in particular stores state and so needs to be stored somewhere rather than being reinstated every time the prompt is sampled.

I agree that the code is hard to read. I wasn't 100% satisfied with it. I think I only properly understood the need for storing state while I was elbows deep in the code.